db/snapshot/backup_task.cc

```diff
@@ -112,39 +116,82 @@ future<> backup_task_impl::do_backup() {
+
+    try {
+        snap_log.debug("backup_task: listing {}", _snapshot_dir.native());
+        for (;;) {
+            auto component_ent = co_await snapshot_dir_lister.get();
+            if (!component_ent.has_value()) {
                 break;
             }
+
+            const auto& name = component_ent->name;
+            try {
+                auto file_path = _snapshot_dir / name;
+                auto [desc, ks, cf] = sstables::parse_path(file_path);
+                sstable_comps[desc.generation].emplace_back(name);
+                ++num_sstable_comps;
+            } catch (const sstables::malformed_sstable_exception&) {
+                non_sstable_files.emplace_back(name);
+            }
         }
-        if (!component_ent.has_value()) {
-            break;
-        }
-        auto gh = uploads.hold();
+        snap_log.debug("backup_task: found {} SSTables consisting of {} component files, and {} non-sstable files",
+                sstable_comps.size(), num_sstable_comps, non_sstable_files.size());
+    } catch (...) {
+        ex = std::current_exception();
+        snap_log.error("backup_task: listing {} failed: {}", _snapshot_dir.native(), ex);
+    }
+    co_await snapshot_dir_lister.close();
+    if (ex) {
+        co_await coroutine::return_exception_ptr(std::move(ex));
+    }
 
-        // Pre-upload break point. For testing abort in actual s3 client usage.
-        co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+    auto backup_comp = [&] (sstring name) {
+        _as.check();
 
-        std::ignore = upload_component(component_ent->name).handle_exception([&ex] (std::exception_ptr e) {
+        auto gh = uploads.hold();
+
+        // okay to drop future since uploads is always closed before exiting the function
+        std::ignore = upload_component(std::move(name)).handle_exception([&ex] (std::exception_ptr e) {
             // keep the first exception
             if (!ex) {
                 ex = std::move(e);
             }
         }).finally([gh = std::move(gh)] {});
-        co_await coroutine::maybe_yield();
-        co_await utils::get_local_injector().inject("backup_task_pause", utils::wait_for_message(std::chrono::minutes(2)));
-        if (impl::_as.abort_requested()) {
-            ex = impl::_as.abort_requested_exception_ptr();
-            break;
+    };
+
+    auto backup_sstable = [&] (sstables::generation_type gen, const comps_vector& comps) -> future<> {
+        snap_log.debug("Backing up SSTable generation {}", gen);
+        for (auto it = comps.begin(); it != comps.end() && !ex; ++it) {
+            // Pre-upload break point. For testing abort in actual s3 client usage.
+            co_await utils::get_local_injector().inject("backup_task_pre_upload", utils::wait_for_message(std::chrono::minutes(2)));
+
+            const auto& name = *it;
+            backup_comp(name);
+
+            co_await coroutine::maybe_yield();
```

Comment:
No need to yield, co_await already does.

Reply:
done in v2 ([06da3f3](https://github.com/scylladb/scylladb/commit/06da3f30f0c2da5e684fbff7b517e97d70d70559))